File: projects/01-ai-knowledge-base/internal/service/audit_service.go
package service

import (
	"context"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"os"
	"time"

	"github.com/SalesChampionHub/ai-knowledge-base/internal/models"
	"github.com/SalesChampionHub/ai-knowledge-base/internal/repository"
)

// AuditService 审计日志服务接口
type AuditService interface {
	// 查询审计日志
	QueryLogs(ctx context.Context, query *models.AuditLogQuery) ([]*models.AuditLog, int64, error)
	
	// 获取审计日志详情
	GetLogByID(ctx context.Context, id int64) (*models.AuditLog, error)
	
	// 获取用户统计
	GetUserStats(ctx context.Context, tenantID, userID string, days int) (*models.UserAuditStats, error)
	
	// 获取租户统计
	GetTenantStats(ctx context.Context, tenantID string, days int) (*models.AuditLogStats, error)
	
	// 导出审计日志
	ExportLogs(ctx context.Context, query *models.AuditLogQuery, format string) (string, error)
	
	// 清理旧日志
	CleanupOldLogs(ctx context.Context, retentionDays int) (int, error)
}

// auditService 审计日志服务实现
type auditService struct {
	auditRepo repository.AuditRepository
}

// NewAuditService 创建审计日志服务
func NewAuditService(auditRepo repository.AuditRepository) AuditService {
	return &auditService{
		auditRepo: auditRepo,
	}
}

// QueryLogs 查询审计日志
func (s *auditService) QueryLogs(ctx context.Context, query *models.AuditLogQuery) ([]*models.AuditLog, int64, error) {
	// 设置默认值
	if query.Page <= 0 {
		query.Page = 1
	}
	if query.PageSize <= 0 {
		query.PageSize = 20
	}
	if query.PageSize > 100 {
		query.PageSize = 100
	}
	
	// 如果没有指定时间范围，默认查询最近30天
	if query.StartTime.IsZero() && query.EndTime.IsZero() {
		query.EndTime = time.Now()
		query.StartTime = query.EndTime.AddDate(0, 0, -30)
	}
	
	return s.auditRepo.Query(ctx, query)
}

// GetLogByID 获取审计日志详情
func (s *auditService) GetLogByID(ctx context.Context, id int64) (*models.AuditLog, error) {
	return s.auditRepo.GetByID(ctx, id)
}

// GetUserStats 获取用户统计
func (s *auditService) GetUserStats(ctx context.Context, tenantID, userID string, days int) (*models.UserAuditStats, error) {
	if days <= 0 {
		days = 30
	}
	return s.auditRepo.GetUserStats(ctx, tenantID, userID, days)
}

// GetTenantStats 获取租户统计
func (s *auditService) GetTenantStats(ctx context.Context, tenantID string, days int) (*models.AuditLogStats, error) {
	if days <= 0 {
		days = 30
	}
	return s.auditRepo.GetTenantStats(ctx, tenantID, days)
}

// ExportLogs 导出审计日志
func (s *auditService) ExportLogs(ctx context.Context, query *models.AuditLogQuery, format string) (string, error) {
	// 获取要导出的日志
	logs, err := s.auditRepo.Export(ctx, query)
	if err != nil {
		return "", err
	}
	
	// 根据格式导出
	switch format {
	case "csv":
		return s.exportToCSV(logs)
	case "json":
		return s.exportToJSON(logs)
	default:
		return "", fmt.Errorf("unsupported export format: %s", format)
	}
}

// exportToCSV 导出为CSV
func (s *auditService) exportToCSV(logs []*models.AuditLog) (string, error) {
	// 创建临时文件
	filename := fmt.Sprintf("/tmp/audit_logs_%d.csv", time.Now().Unix())
	file, err := os.Create(filename)
	if err != nil {
		return "", err
	}
	defer file.Close()
	
	writer := csv.NewWriter(file)
	defer writer.Flush()
	
	// 写入表头
	headers := []string{
		"ID", "租户ID", "用户ID", "用户名", "操作", "资源类型", "资源ID",
		"状态", "IP地址", "请求方法", "请求路径", "耗时(ms)", "创建时间",
	}
	if err := writer.Write(headers); err != nil {
		return "", err
	}
	
	// 写入数据
	for _, log := range logs {
		record := []string{
			fmt.Sprintf("%d", log.ID),
			log.TenantID,
			log.UserID,
			log.Username,
			log.Action,
			log.ResourceType,
			log.ResourceID,
			log.Status,
			log.IPAddress,
			log.RequestMethod,
			log.RequestPath,
			fmt.Sprintf("%d", log.DurationMs),
			log.CreatedAt.Format("2006-01-02 15:04:05"),
		}
		if err := writer.Write(record); err != nil {
			return "", err
		}
	}
	
	return filename, nil
}

// exportToJSON 导出为JSON
func (s *auditService) exportToJSON(logs []*models.AuditLog) (string, error) {
	// 创建临时文件
	filename := fmt.Sprintf("/tmp/audit_logs_%d.json", time.Now().Unix())
	file, err := os.Create(filename)
	if err != nil {
		return "", err
	}
	defer file.Close()
	
	// 编码为JSON
	encoder := json.NewEncoder(file)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(logs); err != nil {
		return "", err
	}
	
	return filename, nil
}

// CleanupOldLogs 清理旧日志
func (s *auditService) CleanupOldLogs(ctx context.Context, retentionDays int) (int, error) {
	if retentionDays <= 0 {
		retentionDays = 365 // 默认保留1年
	}
	return s.auditRepo.CleanupOldLogs(ctx, retentionDays)
}

File: projects/01-ai-knowledge-base-frontend/src/services/auditService.ts
import api from './api';
import type { AuditLog, AuditLogQuery, AuditLogResponse, AuditLogStats, UserAuditStats } from '../types/audit';

/**
 * 审计日志服务
 */
class AuditService {
  /**
   * 查询审计日志
   */
  async queryLogs(query: AuditLogQuery): Promise<AuditLogResponse> {
    const params = new URLSearchParams();
    
    if (query.tenant_id) params.append('tenant_id', query.tenant_id);
    if (query.user_id) params.append('user_id', query.user_id);
    if (query.action) params.append('action', query.action);
    if (query.resource_type) params.append('resource_type', query.resource_type);
    if (query.resource_id) params.append('resource_id', query.resource_id);
    if (query.status) params.append('status', query.status);
    if (query.start_time) params.append('start_time', query.start_time);
    if (query.end_time) params.append('end_time', query.end_time);
    if (query.page) params.append('page', query.page.toString());
    if (query.page_size) params.append('page_size', query.page_size.toString());
    if (query.sort_by) params.append('sort_by', query.sort_by);
    if (query.sort_order) params.append('sort_order', query.sort_order);

    const response = await api.get(`/audit-logs?${params.toString()}`);
    return response.data;
  }

  /**
   * 获取审计日志详情
   */
  async getLogById(id: number): Promise<AuditLog> {
    const response = await api.get(`/audit-logs/${id}`);
    return response.data;
  }

  /**
   * 获取租户统计
   */
  async getTenantStats(days: number = 30): Promise<AuditLogStats> {
    const response = await api.get(`/audit-logs/stats?days=${days}`);
    return response.data;
  }

  /**
   * 获取用户统计
   */
  async getUserStats(userId: string, days: number = 30): Promise<UserAuditStats> {
    const response = await api.get(`/audit-logs/users/${userId}/stats?days=${days}`);
    return response.data;
  }

  /**
   * 导出审计日志
   */
  async exportLogs(query: AuditLogQuery, format: 'csv' | 'json'): Promise<Blob> {
    const params = new URLSearchParams();
    params.append('format', format);
    
    if (query.tenant_id) params.append('tenant_id', query.tenant_id);
    if (query.user_id) params.append('user_id', query.user_id);
    if (query.action) params.append('action', query.action);
    if (query.resource_type) params.append('resource_type', query.resource_type);
    if (query.start_time) params.append('start_time', query.start_time);
    if (query.end_time) params.append('end_time', query.end_time);

    const response = await api.get(`/audit-logs/export?${params.toString()}`, {
      responseType: 'blob',
    });
    return response.data;
  }

  /**
   * 下载导出文件
   */
  downloadExportFile(blob: Blob, filename: string) {
    const url = window.URL.createObjectURL(blob);
    const link = document.createElement('a');
    link.href = url;
    link.download = filename;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    window.URL.revokeObjectURL(url);
  }
}

export default new AuditService();

